engine/grid.py

```diff
@@ -331,36 +337,78 @@ def find_border_nodes(self):
         end of the function call, fields 'leftmost_node', 'leftmost_node_pos', and
         'border_nodes' will be initialized.
         """
+        # initialize variables
         is_vertical = True
         if (
             self.direction == self.Direction.LEFT
             or self.direction == self.Direction.RIGHT
         ):
             is_vertical = False
+
         border_list = []
         leftmost_node = None
-        leftmost_node_pos = None
-
+        bottom_left_pos = None
```

Comment:
alternative solution is to DFS and keep track of bottom left, bottom right, top left, top right. we should also define "bottom_left", "bottom_right", "top_left", and "top_right" for future purposes if we have an irregular shape (think diamond)